### DISCUSSION TITLE
ILL-DMSC Workshop, 10-13 Dec 2024 at ILL / remotely

### DISCUSSION THREAD

**Discussion by AndrewSazonov** [EasyScience/TEST_PyCrysFML#29] [General]
- [ ] Only one combination, macOS 13 (Intel CPU) + gfortran, fails to build the static library libCrysFML08.a, producing the following errors:
  ```
  /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file:   libCrysFML08.a(CFML_Messages.o) has no symbols
  /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file:   libCrysFML08.a(CFML_VTK.o) has no symbols
  /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file:   libCrysFML08.a(KeyCod_RGB.o) has no symbols
  /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file:   libCrysFML08.a(KeyCod_VecRef.o) has no symbols
  ```
- [x] @nebil77: PyCrysFML API documentation https://pycrysfml.readthedocs.io/en/latest/index.html
- [ ] @nebil77: New fortran code for TOF neutron powder diffraction pattern calculation
  - [x] This will be added as a new component of `CFML_Utilities`, in addition to the existing `powder_pattern_from_json` for CW: `cw_powder_pattern_from_json` in the `nebil` branch of CrysFML2008 -> The `master` branch of CrysFML2008
  - [ ] Eventually, both CW and TOF functionalities will be merged into a single, more advanced function
- [x] @nebil77: Modify `powder_pattern_from_json` from CFML_Utilities to follow the CIF (not FullProf) definition of the atomic occupancies: the `nebil` branch of CrysFML2008 -> The `master` branch of CrysFML2008
- [x] @nebil77: Some parameters from the input JSON for CW are not used:
  - [x] `_pd_instr_resolution_y`: We expect to use `u`, `v`, `w`, `x`, and `y` parameters as in FullProf PCR for peak profile `Npr=7`:  the `nebil` branch of CrysFML2008 -> The `master` branch of CrysFML2008
  - [ ] `_pd_instr_reflex_asymmetry_p1`, `_pd_instr_reflex_asymmetry_p2`, `_pd_instr_reflex_asymmetry_p1`, and `_pd_instr_reflex_asymmetry_p4`:  We expect to use `Asy1`, `Asy2`, `Asy3`, and `Asy4` as in FullProf PCR
- [x] `_pd_instr_reflex_s_l`, `_pd_instr_reflex_d_l` are implemented as `S_L`, `D_L` in FullProf PCR -> The `master` branch of CrysFML2008
- [x] @nebil77: There are some minor but noticeable discrepancies between the patterns calculated by FullProf and PyCrysFML. These discrepancies are likely due to differences in the peak profile functions used. This needs to be investigated and resolved. See https://github.com/EasyScience/PyCrysFML/blob/O3/issues/pyCFML/cfml_utilities/powder_pattern_from_json/pycfml_vs_fullprof/comparison.png:  the `nebil` branch of CrysFML2008
- [x] @nebil77: `CFML_Profiles.f90` contains the function `Expi_E1`. If this function is not declared as public using `public :: Expi_E1`, attempting to use it in `CFML_Profiles/Profile_TOF_Jorg_Vondreele.f90` results in the following error when building the PyCrysFML library with gfortran:
  ```
  D:/a/PyCrysFML/PyCrysFML/dist/CFML/lib/CrysFML08.lib(Profile_TOF_Jorg_Vondreele.o):Profile_TOF_Jorg_Vondreele.f90:
  (.text+0x585): undefined reference to `__cfml_profiles_MOD_expi_e1'
  ```
  Temporary solution: clone the branch `nebil` intto `nebil_expi` and add `public :: Expi_E1` to `CFML_Profiles.f90`. However, a permanent solution is needed -> The `master` branch of CrysFML2008
- [ ] @nebil77: There is a difference between FullProf and PyCrysFML in the TOF profile calculated for the Si standard
- [ ] @nebil77: Implement the calculation of the profile for the custom x-array defined in the .json definition of the experiment. For TOF, this is specified under the new key `_pd_meas_time_of_flight`
- [x] @AndrewSazonov: Re-ran the old CI to verify whether the PyCrysFML Python package generation works with the latest master branch of CrysFML2008
  - [ ] New errors on Windows (CFML program builds): `libucrt.lib(frexp.obj) : error LNK2005: frexp already defined in libmmt.lib(frexp_iface_c99.obj)`
- [x] @AndrewSazonov: Meet with JRC to discuss the items mentioned in #10, which are necessary to add support for the `nagfor` compiler in the CrysFML2008 source code
  - [ ] JRC is planning to do this in January 2025, so @AndrewSazonov will contact him after 20/01/2025 to check on progress
- [x] @AndrewSazonov: Check how atomic occupancies are interpreted in PyCrysFML: as in FP or as in CIF?
  - [x] The calculation of the powder diffraction pattern in PyCrysFML takes into account the atomic occupancies from the input json file in the same way as FullProf. That is, one have to manually multiply the occupancy given in CIF by the multiplicity of the current position and divide by the multiplicity of the general position. It would be much simpler if we used the same value in the json as in CIF, since all the other json keys describing the phase have exactly the same meaning as in CIF. CrysFML already has this functionality to read CIF files, so CrysFML should be able to properly handle these values from CIF
- [ ] @AndrewSazonov: Add additional functional tests for PyCrysFML, comparing the results against data calculated by FullProf 
- [ ] @AndrewSazonov: Prepare two examples of single-crystal data analysis using FullProf — one for CW and another for TOF. Send them to Nebil as examples for implementation in PyCrysFML
- [ ] @rozyczko: Add initial CFML (non-functional) wrapper and calculator
  - [x] Add the new crysfml `cfml_utilities.powder_pattern_from_json` call for calculating the profile
  - [X] Run simple (hardcoded) simulation based on the job interface
  - [x] Convert phase to CrysFML json
     - [x]  Modify gemmi cif2json output to fit the json style used by crysfml
  - [X]  Convert experiment to the json style used by crysfml
  - [X] Generate profiles based on experiment/phase cif files (or xye)
  - [X] Generate profiles based on handcoded phase
  - [ ] Assure calculator changes are passing the data properly
  - [ ] Add an example notebook comparing CFML with Cryspy for a simple profile

**Comment by AndrewSazonov** [comment]
- [x] @nebil77: New issue in CrysFML2008/.../gS_Get_Symb_mat.f90. This issue seems to have been introduced by a recent change made by Juan, which has now propagated to the master branch. The error arises from an attempt to construct a CHARACTER array with elements of differing lengths, whereas all elements of a CHARACTER array in Fortran must have the same length, e.g.:
CFML_gSpaceGroups/gS_Get_Orb_Stabilizer_Constr.f90:1211:17:
1211 |         cdw=['a','-a','b']
- [x] @nebil77: There are some minor but noticeable discrepancies between the patterns calculated by FullProf and PyCrysFML for CW. Some spikes appear, which seem to be due to a non-zero S_L parameter.
- [ ] @nebil77: Discrepancy between FullProf and PyCrysFML in generating the TOF diffraction pattern #30 
- [ ] @nebil77: Implement powder diffraction pattern simulation based on an array of x positions for which the diffraction intensities are to be calculated (CW: `_pd_meas_2theta_scan = [...]`, TOF: `_pd_meas_time_of_flight = [...]`)

**Comment by nebil77** [comment]
Spikes were due to a singularity of the convolution function close to 90 degrees. That behavior was protected in CrysFML but not in CrysFML08. Now it is in both.

The problem with gS_Get_Symb_mat has been also solved.
